pkg/reconciler/providerconfig/reconciler.go

```diff
@@ -0,0 +1,198 @@
+	nc := func() resource.ProviderConfig {
+		return resource.MustCreateObject(of.Config, m.GetScheme()).(resource.ProviderConfig)
+	}
+	nul := func() resource.ProviderConfigUsageList {
+		return resource.MustCreateObject(of.UsageList, m.GetScheme()).(resource.ProviderConfigUsageList)
+	}
+
+	// Panic early if we've been asked to reconcile a resource kind that has not
+	// been registered with our controller manager's scheme.
+	_, _ = nc(), nul()
+
+	r := &Reconciler{
+		client: m.GetClient(),
+
+		newConfig:    nc,
+		newUsageList: nul,
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile a ProviderConfig by accounting for the managed resources that are
+// using it, and ensuring it cannot be deleted until it is no longer in use.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	pc := r.newConfig()
+	if err := r.client.Get(ctx, req.NamespacedName, pc); err != nil {
+		// In case object is not found, most likely the object was deleted and
+		// then disappeared while the event was in the processing queue. We
+		// don't need to take any action in that case.
+		log.Debug(errGetPC, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPC)
+	}
+
+	log = log.WithValues(
+		"uid", pc.GetUID(),
+		"version", pc.GetResourceVersion(),
+		"name", pc.GetName(),
+	)
+
+	l := r.newUsageList()
+	if err := r.client.List(ctx, l, client.MatchingLabels{v1alpha1.LabelKeyProviderName: pc.GetName()}); err != nil {
+		log.Debug(errListPCUs, "error", err)
+		r.record.Event(pc, event.Warning(reasonAccount, errors.Wrap(err, errListPCUs)))
+		return reconcile.Result{RequeueAfter: shortWait}, nil
+	}
+
+	users := int64(len(l.GetItems()))
+	for _, pcu := range l.GetItems() {
+		if metav1.GetControllerOf(pcu) == nil {
+			// Usages should always have a controller reference. If this one has
+			// none it's probably been stripped off (e.g. by a Velero restore).
+			// We can safely delete it - it's either stale, or will be recreated
+			// next time the relevant managed resource connects.
+			if err := r.client.Delete(ctx, pcu); resource.IgnoreNotFound(err) != nil {
+				log.Debug(errDeletePCU, "error", err)
+				r.record.Event(pc, event.Warning(reasonAccount, errors.Wrap(err, errDeletePCU)))
+				return reconcile.Result{RequeueAfter: shortWait}, nil
+			}
+			users--
+		}
+	}
+	log = log.WithValues("usages", users)
+
+	if meta.WasDeleted(pc) {
+		if users > 0 {
+			log.Debug("Blocking deletion while usages still exist")
+			r.record.Event(pc, event.Warning(reasonAccount, errors.New("Blocking deletion while usages still exist")))
+
+			// We're watching our usages, so we'll be requeued when they go.
+			pc.SetUsers(&users)
```

Comment:
Should we set a condition, too? Something like `Terminating`.

Reply:
We could, though we'd presumably need some condition that implied something like "not terminating, but not really confirmed to actually be ready/working/etc". Maybe `Active`?

Reply:
`Active`, `Used` or `Connected` all sound good to me.

Reply:
SGTM - I'll probably go with `Active`. I don't like `Connected` in particular, because it could be interpreted as resources have successfully "connected" to the cloud using this `ProviderConfig`, which we don't guarantee.

Reply:
In retrospect, we don't need any inverse condition. It's fine for the `Terminating` condition to appear only when the PC is being deleted - it will never transition out of that condition.